Change the software interrupt to syscall instruction

chirping78 opened the pull request:
## Summary

Currently xtensa arch porting is using software interrupt (through write `intset` register) to trigger context switch.
But this put a strict requirement to the chip configuration: 
there must be a software interrupt, it's priority is less or equal to XCHAL_EXCM_LEVEL;
and it's priority must be greater than XCHAL_SYSCALL_LEVEL (the XCHAL_SYSCALL_LEVEL is not a cadence macro, it just means normal hardware interrupt priority).

Since the xtensa arch is very flexible, the chip designer might have not noticed above requirement in chip design phase.
And if the chip has been taped out without this requirement, there is no posibility to port nuttx to the chip.

This PR will try to change the software interrupt to `syscall` instruction.
The `syscall` instruction is a synchronous exception, and is not locked by `PS.intlevel`, which means it's priority is high enough.

## Impact

This PR will impact context switch, interrupt handling, and protect build.

## Testing

test summary 

12/18: rebased to nuttx  d15a20eac85, apps   fa22f80ee.
This time all below configs ostest passed.

esp32s3-devkit:nsh ostest 
esp32s3-devkit:tickless ostest 
esp32s3-devkit:knsh ostest 
esp32s3-devkit:smp ostest 

12/02: rebased to nuttx 140b3080c5f, apps 296d5d15a. 
This time only smp failed at "nested signal handler test". Other config passed.

esp32s3-devkit:nsh       ostest pass
esp32s3-devkit:tickless  ostest pass
esp32s3-devkit:knsh      ostest pass

esp32s3-devkit:smp ostest failed at "nested signal handler test"

11/18: rebased, better than last time.

rebase to version: nuttx master 637a0f5af23, apps master fee82bd3d,
plus this PR.

esp32s3-devkit:nsh       ostest pass
 
esp32s3-devkit:tickless  ostest pass

esp32s3-devkit:smp       ostest pass with percentage ~20% 
The possibility of passing the test is higher if power off then on through plug and unplug usb.

esp32s3-devkit:knsh      ostest pass with percentage ~40%

11/13: 

With https://github.com/apache/nuttx/pull/14760  esp32s3-devkit:smp ostest is more stable, but still not 100% pass.

11/05: has regression problem

base version: nuttx master 56f292fcfa9, apps  master 400a8e326,
plus this PR.

esp32s3-devkit:nsh (+CONFIG_TESTING_OSTEST)
 ostest pass

esp32s3-devkit:tickless
 ostest pass

esp32s3-devkit:smp
 ostest failed (only "nested signal handler test" failed)

esp32s3-devkit:knsh
 ostest failed (fail at random position with message "xtensa_panic: Unhandled Exception 2", 2 means double exception)

Since there is no esp32 and esp32s2 boards in hand (have placed order), so these boards have not been tested.

xiaoxiang781216 commented:
@tmedicci could you help review this pr?

tmedicci commented:
I'm busy during these days, so I'll take a while to fully review it. For now, I'll run our internal pipeline and report its results.

But I have question: IINM, `syscall` was being used for context switch and it was reverted to the software interrupt. I don't recall the context of the discussion, but it'd be good to try to retrieve it in this PR.

(@Ouss4 and @gustavonihei may have some thoughts about it. If they have spare time, I'd like to hear about...).

tmedicci commented:
Hi @xiaoxiang781216 , can we mark it as a draft until we fully review this PR to avoid it being merged by mistake?

tmedicci commented:
> For now, I'll run our internal pipeline and report its results.

The CI shows the same failed results regarding `ostest` for Xtensa devices.

tmedicci commented:
> > For now, I'll run our internal pipeline and report its results.
> 
> The CI shows the same failed results regarding `ostest` for Xtensa devices.

Hi @chirping78 , were you able to verify why `ostest` fails?

chirping78 commented:
> 
> Hi @chirping78 , were you able to verify why `ostest` fails?

@tmedicci 
I'm still investigating these failing cases, and have not root caused them.
Any help or patch is welcome.

chirping78 commented:
With https://github.com/apache/nuttx/pull/14760  esp32s3-devkit:smp ostest is more stable, but still not 100% pass.

tmedicci commented:
> With #14760 esp32s3-devkit:smp ostest is more stable, but still not 100% pass.

There were some changes recently that affect `ostest` behavior, can you please rebase your code?

chirping78 commented:
12/02: rebased to nuttx 140b3080c5f, apps 296d5d15a. 
This time only smp failed at "nested signal handler test". Other config passed.

esp32s3-devkit:nsh       ostest pass
esp32s3-devkit:tickless  ostest pass
esp32s3-devkit:knsh      ostest pass

esp32s3-devkit:smp ostest failed at "nested signal handler test"

tmedicci commented:
> esp32s3-devkit:smp ostest failed at "nested signal handler test"

Isn't it related to https://github.com/apache/nuttx/issues/15059?

chirping78 commented:
> > esp32s3-devkit:smp ostest failed at "nested signal handler test"
> 
> Isn't it related to #15059?

I have downgraded my local toolchain, and then rebased to 12/15 (nuttx d15a20eac85, apps  fa22f80ee).
Now met a new problem in `up_exit` / `xtensa_context_restore`, which means some changes between nuttx(140b3080c5, d15a20eac85) is not compatiable with this PR.
This will take some time to investigate why?

chirping78 commented:
12/18: rebased to nuttx  d15a20eac85, apps   fa22f80ee.
This time all below configs ostest passed.

esp32s3-devkit:nsh ostest pass
esp32s3-devkit:tickless ostest pass
esp32s3-devkit:knsh ostest pass
esp32s3-devkit:smp ostest pass

xiaoxiang781216 commented:
@tmedicci please review and verify the change work with your internal test.

xiaoxiang781216 commented on arch/xtensa/src/common/xtensa_int_handlers.S:
```diff
@@ -197,6 +197,13 @@ _xtensa_level1_handler:
+
+	movi	a2, ~PS_EXCM_MASK
+	l32i	a3, sp, (4 * REG_PS)
+	and		a3, a3, a2
```
```
and	a3, a3, a2
```

xiaoxiang781216 commented on arch/xtensa/src/common/xtensa_int_handlers.S:
```diff
@@ -218,6 +225,13 @@ _xtensa_level1_handler:
+
+	movi	a3, PS_EXCM_MASK
+	l32i	a4, a2, (4 * REG_PS)
+	or		a4, a3, a4
```
```
or	a4, a3, a4
```

xiaoxiang781216 commented on arch/xtensa/src/common/xtensa_user_handler.S:
```diff
@@ -193,6 +193,13 @@ _xtensa_user_handler:
+
+	movi	a3, PS_EXCM_MASK
+	l32i	a4, a2, (4 * REG_PS)
+	or		a4, a3, a4
```
ditto

xiaoxiang781216 commented on arch/xtensa/src/common/xtensa_user_handler.S:
```diff
@@ -235,6 +242,13 @@ _xtensa_syscall_handler:
+
+	movi	a2, ~PS_EXCM_MASK
+	l32i	a3, sp, (4 * REG_PS)
+	and		a3, a3, a2
```
ditto

xiaoxiang781216 commented on arch/xtensa/src/common/xtensa_user_handler.S:
```diff
@@ -304,6 +318,13 @@ _xtensa_syscall_handler:
+
+	movi	a3, PS_EXCM_MASK
+	l32i	a4, a2, (4 * REG_PS)
+	or		a4, a3, a4
```
ditto

tmedicci commented:
> @tmedicci please review and verify the change work with your internal test.

Triggering the internal CI. I'll share the results soon. Please don't merge it yet.

tmedicci commented on arch/xtensa/src/common/xtensa_int_handlers.S:
```diff
@@ -197,6 +197,13 @@ _xtensa_level1_handler:
 
 	call0	_xtensa_context_save
 
+	/* Overwrite the PS saved by the exception entry. */
```
This is clearing the PS.EXCM, right? If so, why?

`CINTLEVEL ← max(PS.EXCM∗EXCMLEVEL,PS.INTLEVEL)`

So, clearing the PS.EXCM would enable other interrupts to be served, but NuttX doesn't support nested interrupts.

chirping78 replied:
This clearing the saved PS.EXCM, not current PS.
I have added the explain here:
https://github.com/apache/nuttx/pull/14672#issuecomment-2553126081

tmedicci commented on arch/xtensa/src/common/xtensa_user_handler.S:
```diff
@@ -193,6 +193,13 @@ _xtensa_user_handler:
 
 	mov		a2, RETVAL							/* a2 = address of new state save area */
 
+	/* Set PS.EXCM to 1 */
+
+	movi	a3, PS_EXCM_MASK
+	l32i	a4, a2, (4 * REG_PS)
+	or	a4, a3, a4
+	s32i	a4, a2, (4 * REG_PS)
```
Shouldn't it be called before [`xtensa_user`](https://github.com/apache/nuttx/pull/14672/files#diff-bf4f09df2936401ec123284e73fc661258f077f216afde7bc32d2ba5706ff831R192)?

chirping78 replied:
Should be here. I have added explain here: https://github.com/apache/nuttx/pull/14672#issuecomment-2553126081

tmedicci commented:
> > @tmedicci please review and verify the change work with your internal test.
> 
> Triggering the internal CI. I'll share the results soon. Please don't merge it yet.

It isn't building any of the ESP32-S2's defconfig. I tried locally and found the same issue:

```
make -j distclean && ./tools/configure.sh esp32s2-saola-1:ostest && make olddefconfig && make flash ESPTOOL_PORT=/dev/ttyUSB0 EXTRAFLAGS="-Wno-cpp -Werror" ESPTOOL_BINDIR=./ -s -j$(nproc) &&  xtensa-esp32s2-elf-objdump -D nuttx  > objdump.flat && minicom -D /dev/ttyUSB0
  Copy files
  Select CONFIG_HOST_LINUX=y
  Refreshing...
CP: arch/dummy/Kconfig to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/dummy/dummy_kconfig
CP: boards/dummy/Kconfig to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/dummy/dummy_kconfig
LN: platform/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/platform/dummy
LN: include/arch to arch/xtensa/include
LN: include/arch/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/include
LN: drivers/platform to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/drivers/dummy
LN: include/arch/chip to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/include/esp32s2
LN: arch/xtensa/src/chip to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/esp32s2
LN: arch/xtensa/src/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common
LN: arch/xtensa/src/board/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/src
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/audioutils
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/benchmarks
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/boot
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/canutils
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/crypto
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/database
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/examples/mcuboot
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/examples/module
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/examples/sotest
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/examples
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/esp-nuttx-apps/examples
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/esp-nuttx-apps/testing
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/esp-nuttx-apps
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/fsutils
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/games
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/graphics
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/industry
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/inertial
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/interpreters/luamodules
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/interpreters
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/logging
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/lte
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/math
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/mlearning
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/netutils
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/sdr
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/system
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/testing
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/videoutils
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/wireless/bluetooth
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/wireless/ieee802154
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/wireless
mkkconfig in /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps
Loaded configuration '.config'
Configuration saved to '.config'
CP: arch/dummy/Kconfig to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/dummy/dummy_kconfig
CP: boards/dummy/Kconfig to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/dummy/dummy_kconfig
LN: platform/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/platform/dummy
LN: include/arch to arch/xtensa/include
LN: include/arch/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/include
LN: drivers/platform to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/drivers/dummy
LN: include/arch/chip to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/include/esp32s2
LN: arch/xtensa/src/chip to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/esp32s2
LN: arch/xtensa/src/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common
LN: arch/xtensa/src/board/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/src
Loaded configuration '.config'
No change to configuration in '.config'
Create version.h
Cloning Espressif HAL for 3rd Party Platforms
Clone: chip/esp-hal-3rdparty LN: platform/board to /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/apps/platform/dummy
Register: nsh
Register: sh
Register: ostest
Espressif HAL for 3rd Party Platforms: e3899a2324c8e326db20f99f208e890fdd7a5b92
CPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common/scripts/esp32s2_aliases.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boardCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common/scripts/esp32s2_peripherals.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/bCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common/scripts/flat_memory.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/boards/xtensa/esp32s2/esp32s2-saola-1/../common/scripts/simple_boot_sections.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/CPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.api.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_betaCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nutCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.libgcc.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_bCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-data.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-funcs.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxsCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/esp_rom/esp32s2/ld/esp32s2.rom.spiflash.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspaceCPP:  /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_beta/nuttx/arch/xtensa/src/chip/esp-hal-3rdparty/components/soc/esp32s2/ld/esp32s2.peripherals.ld-> /home/tiago/Documents/work/espressif/projects/nuttx/nuttxspace_betaLD: nuttx
CP: nuttx.hex
MKIMAGE: ESP32-S2 binary
esptool.py v4.8.dev4
Creating esp32s2 image...
Image has only RAM segments visible. ROM segments are hidden and SHA256 digest is not appended.
Merged 1 ELF section
Successfully created esp32s2 image.
Generated: nuttx.bin
esptool.py v4.8.dev4
Serial port /dev/ttyUSB0
Connecting......
Chip is ESP32-S2 (revision v0.0)
Features: WiFi, No Embedded Flash, No Embedded PSRAM, ADC and temperature sensor calibration in BLK2 of efuse V2
Crystal is 40MHz
MAC: 60:55:f9:ed:a2:92
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 4MB
Flash will be erased from 0x00001000 to 0x00038fff...
Compressed 229064 bytes to 99972...
Wrote 229064 bytes (99972 compressed) at 0x00001000 in 3.4 seconds (effective 541.7 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...


Welcome to minicom 2.9

OPTIONS: I18n 
Compiled on May  3 2024, 16:19:02.
Port /dev/ttyUSB0, 11:10:40

Press CTRL-A Z for help on special keys

ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:2
load:0x3ffb8760,len:0x56c
load:0x40022000,len:0x4084
SHA-256 comparison failed:
Calculated: 8aaa6fc20ad0f20f1499127b2b8fbdd004660e05664391fbd2709124a8a45797
Expected: 00000000d0a90000000000000000000000000000000000000000000000000000
Attempting to boot anyway...
entry 0x40022930
*** Booting NuttX ***
dram: lma 0x00001020 vma 0x3ffb8760 len 0x56c    (1388)
iram: lma 0x00001594 vma 0x40022000 len 0x4084   (16516)
padd: lma 0x00005628 vma 0x00000000 len 0xa9d0   (43472)
imap: lma 0x00010000 vma 0x40090000 len 0x1983c  (104508)
padd: lma 0x00029844 vma 0x00000000 len 0x67d4   (26580)
dmap: lma 0x00030020 vma 0x3f000020 len 0x8ea8   (36520)
total segments stored 6
xtensa_panic: Unhandled Exception 2 task: Idle_Task
dump_assert_info: Current Version: NuttX  10.4.0 696c7c00d1 Dec 18 2024 11:10:21 xtensa
dump_assert_info: Assertion failed panic: at file: :0 task: Idle_Task process: Kernel 0x40091c50
up_dump_register:    PC: 400223df    PS: 00060936
up_dump_register:    A0: 00000009    A1: 3fffffb5    A2: ffffffef    A3: 00060922
up_dump_register:    A4: 3ffb61cc    A5: 00060920    A6: 3ffb6150    A7: 0000000c
up_dump_register:    A8: 00000000    A9: 3fffe4d0   A10: 00000001   A11: 00000001
up_dump_register:   A12: 00000000   A13: 00000000   A14: 3ffb8fa8   A15: 00000000
up_dump_register:   SAR: 00000020 CAUSE: 00000009 VADDR: 40000001
dump_stacks: ERROR: Stack pointer is not within the stack
dump_stackinfo: User Stack:
dump_stackinfo:   base: 0x3ffb7b70
dump_stackinfo:   size: 00003056
stack_dump: 0x3ffb7b70: d0cf5901 f4a090fd 4429b627 5f073b4f a3b4ff28 4099bf7a 409e5d0a bcdb2fef
stack_dump: 0x3ffb7b90: a9f50971 5968a79c 170a531f aadbd89e d753a9b7 22c31139 48b3b183 6cea4557
stack_dump: 0x3ffb7bb0: cf739970 b449b6d1 3d20a69c be5f2833 a314ac08 61f386af 5b8cef98 2a565cd8
stack_dump: 0x3ffb7bd0: acb877e5 b83f6245 2ebb5cb6 dbb2dc9d 95029ece 724d697b 83d43a0d d5994879
stack_dump: 0x3ffb7bf0: 3c7d239a 770074ae 2eb692b1 386d3b08 33613d4d 6b4da1d4 91fe0d23 04e7c6c4
stack_dump: 0x3ffb7c10: 86416f5a f76d4fa3 2edeb49c f9fd35a3 f59ff99b 01b377ea 036e179e 505ecf96
stack_dump: 0x3ffb7c30: b70141cc 72b0f4d1 b90d9528 cabf661c e718190e 75d8e283 0cfd2c94 48d46ea2
stack_dump: 0x3ffb7c50: 5b6e7684 10693878 d5dff632 a5d8adfd 3ff2b10f 63e6439a 245100c0 ce4e506d
stack_dump: 0x3ffb7c70: dfd837cc 758e22d8 723c2cf2 dcca90dd bf6c1ecf 7b82b0d7 443cc6de f3a79a5d
stack_dump: 0x3ffb7c90: b294f133 cd40ff52 6eeef8bc 75fe7c42 82f56e23 5bfbb512 f2e6833d 83ce10f4
stack_dump: 0x3ffb7cb0: d325fc62 0b1c4faf 84178674 168ec5fe 494e2833 4c6bc55a c1575cde b53695e7
stack_dump: 0x3ffb7cd0: 13d47469 be09db9e 5d530e14 1f94c5be 358e067e 25c2fd2c a5519347 b2b1af84
stack_dump: 0x3ffb7cf0: 8a76913a b2738258 a61af345 7f0cecdf 7b2d0aa9 98aa5322 465829a5 32454f13
stack_dump: 0x3ffb7d10: 2acdee3f af2291f5 6d92bbef 195c80d9 ed1db8e5 26951e0a 6e992114 084ec1c2
stack_dump: 0x3ffb7d30: aabcf3e1 f52361fb 5b14c656 4402d181 0f9b8ca7 9efed101 e2442322 8e260bcb
stack_dump: 0x3ffb7d50: 7f49cd7c 09a76fd6 9a06b072 2a851525 e420bc9a efa5561a cd7018a5 5a716b95
stack_dump: 0x3ffb7d70: 4c8c67b3 e488cc87 ff6f6feb 4fff5800 be8d4ecc 3fc13a96 e7d8a08b b834b8f4
stack_dump: 0x3ffb7d90: 94700ab7 dbb5dd32 99d70a65 b9265b9e bbbb293b ba7fa74a 1cb4d767 2ba67dcf
stack_dump: 0x3ffb7db0: 74cf6308 0ebb3672 de30705c fcfc6bbe 945ed42b d301d53a 2623e823 92b728de
stack_dump: 0x3ffb7dd0: 7fa85342 dfe9bebf 75ed1d22 79858fac c5457907 9a1e9c0a 1e86ce4f 5adc32ce
stack_dump: 0x3ffb7df0: 5c56f187 7588dff7 3a9de778 4067cd8c 07791c53 a3bd7c61 12f9f5b0 0722f1f4
stack_dump: 0x3ffb7e10: 156f6686 e5545bfa 1907ad9e 4831d85e 2d7c9ff7 de98e8e8 fa77fbba 56d8de85
stack_dump: 0x3ffb7e30: f9e6f2c7 5700744b 9933bfbc e2b1ccca 790bbe11 cbb8ca4a 1f96e7a1 67f9d560
stack_dump: 0x3ffb7e50: 3eea39c5 0170bf6e 7c4024d8 6eab14c9 e6d3b262 07414769 314046cb b3117b96
stack_dump: 0x3ffb7e70: b1b8ab4d 499a7566 35b20a0a bb1c18be 9eaac0f9 5c789a8d 19ecce9a 85c53798
stack_dump: 0x3ffb7e90: 54a12e03 77be7367 50321091 9931ac98 fa06b5a6 ddc1ce7c b7f7a8f7 3d123099
stack_dump: 0x3ffb7eb0: 905c027f 45614985 94fc58c0 e91b0447 e50fd8bc 84d12e38 8d6d3c92 499ac07e
stack_dump: 0x3ffb7ed0: db9ab7cc 95066a66 6dd1a65f 8ef98d55 1ae4be70 2ed06851 da1475a4 ed107ce1
stack_dump: 0x3ffb7ef0: d672efd7 ae6bcb36 875f9b2f 7fc9c451 263f3fef 3e6b75a9 7bd78493 c7273d14
stack_dump: 0x3ffb7f10: 3fa1c54d 8dc0c75a 7e3dafac 13f737d0 9177fc96 8505fcef 6f2fecb3 03a16f91
stack_dump: 0x3ffb7f30: 988c90ae 248cdb01 675ad377 862aa1da fcd4bd55 fb9ddfaf df110545 309bb94f
stack_dump: 0x3ffb7f50: c8f8805a b1050edc c37db7d3 e92ca006 a7b4fc0a dd1ab944 44840ee5 387fb400
stack_dump: 0x3ffb7f70: ca889f50 fab58a95 0e916778 d12b8b7b 7b9066fb 3188947c fe90cb14 268ec463
stack_dump: 0x3ffb7f90: 5e61fb8c 2c006417 d3438389 40efd9cc e4c7be0f e03ddfd5 509aeeb3 0b36be2e
stack_dump: 0x3ffb7fb0: 22e6b7e2 2c86b5d5 0fb268be ccaec72e 54f18edd d149cdee 11af8cbe 5d01e935
stack_dump: 0x3ffb7fd0: cb3c1f2d 0f04e464 9b094b68 07ae2efe 98499d4c 78d6d802 3a7d3d99 7dac5fd7
stack_dump: 0x3ffb7ff0: dfa9475a b13c02d9 bd152570 042ce42c bdaada6c 05906f6b 36bb3d56 bc378409
stack_dump: 0x3ffb8010: e4023932 76e52c01 f4da3c24 644a118d f77ed6fa 5f9f58f8 ad2fb5e5 4968fb99
stack_dump: 0x3ffb8030: f22e91fe 7d7e3ce2 2bf3f2ac a87bab4d d0b5c046 f6964c5e aca51974 f89d4d74
stack_dump: 0x3ffb8050: 051660f7 f85f350c 6640d98f 7fcdbe64 a15630bb a742be50 d9d37efa 51fa85b1
stack_dump: 0x3ffb8070: c9656898 7784f1ae ee048013 9f59e48f d66bc81c f63a3706 b748c5c6 657679b1
stack_dump: 0x3ffb8090: dad2798a 7ab2844b 58c56d59 45bbf0aa ce43a7b4 e901d58a d6d2dd8a 7341f27d
stack_dump: 0x3ffb80b0: fb3e9544 9c93f4ce f31496d4 2a181dbd da7a2215 4ae111a3 9bc94009 270ff10c
stack_dump: 0x3ffb80d0: 05c4a74d d16d55f7 c0d594bd 4b8b7d9e 218f3092 48eddac9 db8f34bd 10b6602d
stack_dump: 0x3ffb80f0: 746c8c31 c900f914 5e7e0fec d0f249b9 99a78d6e 65b9cadc 6a98a5df dacc83f2
stack_dump: 0x3ffb8110: b03d22c5 39f1d0f4 cc920b44 e4601aa9 12b9f063 8793abdf 0830bed0 9a459406
stack_dump: 0x3ffb8130: 4475ad8d df5977a2 95466a77 e70bcfce a2b4c33d 136413fb b1fb1c35 97ef1485
stack_dump: 0x3ffb8150: d8d8cdca 4062ef86 a25db888 263f7976 fd5b735c 61c51a2b a75574f2 6ff3ca5d
stack_dump: 0x3ffb8170: 8311fe3a fd7584cd 3c82d01d 0482a4d8 fa5d3d28 6bdeca73 344d99c9 8250a754
stack_dump: 0x3ffb8190: a4b2882f b756760d 9e03691a 726feefe 0cb61e16 1d84e665 ec21b388 7a340190
stack_dump: 0x3ffb81b0: 5dd9e0e3 b11f9486 c19364de cb65aeba 61244c2e 20dc6ea1 eac739c3 fee2fea8
stack_dump: 0x3ffb81d0: 1ee98c02 13eabd7e 424a7ee2 ac95b196 afd51a4c bf26291f 15f46f92 a5864484
stack_dump: 0x3ffb81f0: 2035eb20 885d7845 ee99f921 35e0bb5b b911bd36 7b280b5c 9d6cc5b4 c0661bae
stack_dump: 0x3ffb8210: 4697c42e 82c614e8 0f66f016 ddeaf89a c5700a66 68f041f3 abe55fb0 bf89bc6a
stack_dump: 0x3ffb8230: 9e227ea0 2f8843a4 1e7472d9 4d204109 54c5a813 c931380d 2d940769 a2819abb
stack_dump: 0x3ffb8250: 98676889 a0300d6f f0b02718 b66bf736 157b6d53 1e72105a 7ff61109 f7e56a57
stack_dump: 0x3ffb8270: feb31a90 756b4f38 68474740 6e88f36b 3eebbac4 9fa78d48 715eb541 7328bad1
stack_dump: 0x3ffb8290: e0fd5a08 c1badb2f ad7ae732 f3b81b5f 5767b0e5 bb0790e8 264da52d 9a33f34b
stack_dump: 0x3ffb82b0: 10d9c9d0 8f4bf7a8 87416cfc 73033b26 2b6cf49b 0ab99aa0 cf0d5769 a8ff0f41
stack_dump: 0x3ffb82d0: 057480a1 bc5210e1 69e3fef2 99eefe76 c9821dfd 89ae79c5 c344e0bb 4ca38361
stack_dump: 0x3ffb82f0: 442eb2ae a3b38cf0 b47f589c 61bf598e 0ac83fe6 82763375 041fccff ac1c703b
stack_dump: 0x3ffb8310: f1494e3d 13ce25b0 f7fbb913 7d9758e5 f538258f d45a4134 e92681ac 4520e93d
stack_dump: 0x3ffb8330: 3662b4d5 6e90fb1b 6ffa6831 81d0ab30 7b5d0df6 d1c81c1e 7c2cce82 57441eac
stack_dump: 0x3ffb8350: 57c96b07 d50ae8c1 afcec5cc aab7ff1e 380c4cd2 05e7d7cf 088f693c e395e33e
stack_dump: 0x3ffb8370: ddfb8ccf b06673c0 3ef3c2f7 0be38f02 daf99775 d5adb73f f1ad30c2 48442f83
stack_dump: 0x3ffb8390: 2a61e8ec c534fb12 ef42a901 10eb5bda dc37abff 1c88f47f d9fac953 0dd48b9a
stack_dump: 0x3ffb83b0: 26cc9baf 1b8b6a67 93a7a002 e2973eee 7dca128d 7ae81685 3b207878 94a890ef
stack_dump: 0x3ffb83d0: 7611e88b 24aa159a 16a5474e 1950bed0 d6aa117b a28a8adf 5bbd6db8 7e185665
stack_dump: 0x3ffb83f0: a376b838 297eff6c 0fefcd0c 2ccaac0f ab0234bd 709f3f64 81ddb8a7 557ebce9
stack_dump: 0x3ffb8410: ded3bc32 ea2b1d18 ca0861f4 33dfe75a fd91a671 81716581 225d4381 a32bb175
stack_dump: 0x3ffb8430: 851ab783 813b3fd7 f2fa73de e1396ca1 25a1356a 25b96911 ce911cb9 2e0ae495
stack_dump: 0x3ffb8450: 0e87096f d2009d15 591b352d 209efbb3 f52bc4f8 55fd1f23 13959834 d4863a3a
stack_dump: 0x3ffb8470: a2a157b1 70a8ff13 b8fdb59b d11d6c9f e1ef095d 55a80f49 91dfa460 eb93dd46
stack_dump: 0x3ffb8490: fa7b2407 ef11dd56 b8ef9d50 8b0b3f16 a770311a fe922c11 d72a01cc 45ff2d27
stack_dump: 0x3ffb84b0: 1a5c2014 4181c965 1b5a0880 ede5ab13 2e23c554 9aa0b311 f8cce2ce 368e9822
stack_dump: 0x3ffb84d0: aed57cf5 7009fd21 f00d1e37 66b9eaab b5b850ed 7346e1b0 fea2b88e 4c44815b
stack_dump: 0x3ffb84f0: a19d2c44 1d9b658e 637d2710 9651b86f 535ccef1 6d4b8e30 2bbb7e2c f1ef9a92
stack_dump: 0x3ffb8510: 3adc9592 3fe0f0ae ce110512 a2387465 43af32fc 13d2caca 019a2521 04399ea6
stack_dump: 0x3ffb8530: efd8775b f4cbdafe d59b9be3 50529390 9045d9f7 75db7a75 9087dde0 6625bbdc
stack_dump: 0x3ffb8550: c10e887a e8a65ac8 8d8c4892 5e4728b7 fb88078e cebdab6e 009e5556 8936a576
stack_dump: 0x3ffb8570: a4a3744b a8598e5a ab32595e 118be28b b8c10439 1cfb77bb 2176925e 653bb8cf
stack_dump: 0x3ffb8590: b66330ed 18e5ed6b 583902a2 eaaf8b6e 32b08c4a 491ffc1c ec697e8e 3f19eb0d
stack_dump: 0x3ffb85b0: 9ab97412 2737250b dede2b81 2fd3ae24 f37e740e 672c3dfd 22946ab2 c98c7e3c
stack_dump: 0x3ffb85d0: ba1bf19b 006ff351 6f8eda15 c7f06f6b b6d01a4e f4f43925 37f9c923 271c47d8
stack_dump: 0x3ffb85f0: dd62e836 2e572c7b 5fa6ead2 9d7101a8 0df591d4 d52eb3fd d60922d6 552dff84
stack_dump: 0x3ffb8610: 885112b0 f3ed20af bb08a198 a1267348 84f785cb a62ba273 0c0b6233 687d9196
stack_dump: 0x3ffb8630: e41ef0cc 011665be 74ec73a5 4eed3eae 5f8e14c5 2802e587 77ba65b6 ce1bbbf7
stack_dump: 0x3ffb8650: f8ad74ec 46c49b15 94bcd435 da0bbe63 7b6d45ef 43b261fb a5ee88a9 909abcea
stack_dump: 0x3ffb8670: ec5d1de8 e68c71b6 cf35059e 9b53b67e a83bd012 50aedcc1 30125040 18c412ba
stack_dump: 0x3ffb8690: 94019700 bf72536d 4688d029 f48dcae6 a64aea9c 92e99dcb cda95809 baa60650
stack_dump: 0x3ffb86b0: 887a3d05 6154b55d 57732fb1 79b6f867 46ff6e6d a6348989 e203ccb1 3140869d
stack_dump: 0x3ffb86d0: 20eff2b4 d0e6b64c e9d0fa53 e54bb88b 740ad643 01ef4c83 43973a10 56c236cc
stack_dump: 0x3ffb86f0: ca2b8b80 f85c1c70 a8ece512 97607339 465806c1 8d82b773 97b59c8b 5a7e0ea9
stack_dump: 0x3ffb8710: 0d4c726d 1984d660 668bad4a 6060bfbd d7636081 74ce808b a83696df 692bd9ef
stack_dump: 0x3ffb8730: 8eade8b6 14398426 5f5bf947 b6a095ed 87d9d014 aa2cee7c edba689a 424aa12e
stack_dump: 0x3ffb8750: 51b63ecc 06c0b752 3c6f8dde 6b385583 00000000 00000000 00000000 00000000
dump_tasks:    PID GROUP PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE   COMMAND
dump_task:       0     0   0 FIFO     Kthread -   Ready              0000000000000000 0x3ffb7b70      3056   Idle_Task
dump_task:       1     1 100 RR       Task    -   Running            0000000000000000 0x3ffb93f8      3016   nsh_main
```

I tried this branch without any modifications with https://github.com/apache/nuttx-apps/commit/2e0b5776804547bfa4df5bd5b34627b84c04475d

chirping78 commented:
Will try locally on ESP32-S2 to see whether can reproduce the issue @tmedicci

chirping78 commented:
@tmedicci 
As you can see in this PR, each time when enter level1 interrupt/exception, and after `_xtensa_context_save`,
I added this assemble snippet:
```
+       /* Overwrite the PS saved by the exception entry. */
+
+       movi    a2, ~PS_EXCM_MASK
+       l32i    a3, sp, (4 * REG_PS)
+       and     a3, a3, a2
+       s32i    a3, sp, (4 * REG_PS)
+
```
Before leave from level1 interrupt/exception, and before `_xtensa_context_restore`,
the following snippet is added:
```
+       /* Set PS.EXCM to 1 */
+
+       movi    a3, PS_EXCM_MASK
+       l32i    a4, a2, (4 * REG_PS)
+       or      a4, a3, a4
+       s32i    a4, a2, (4 * REG_PS)
+
```


Here is the explaination of why these two snippets are added.

As been defined by Xtensa Architecture, the level1 interrupt is different from other interrupt levels:

1. the level1 interrupt has no it's own `EPS_X` register to save `PS`. 
There are only `EPS2..7` for interrupt above level1. This means the `REG_PS` saved in the task stack is not `PS` value before interrupt.

2. the instruction to return from level1 interrupt is `RFE`, while `RFI` for other interrupt levels.
`RFE` sets `PS.EXCM` back to 0, while `RFI` restores the `PS` from `EPS_X`.

Taking into consideration the above two facts, the level1 interrupt needs special handling:
- after `_xtensa_context_save`, `EXCM` bit needs be cleard from `REG_PS`, since we don't know the interruptee task will be resumed by `RFE` or `RFI`
- before `_xtensa_context_restore`, `EXCM` bit needs be set to `REG_PS`, or there will be risky to interrupt nested in `exception_exit`/`ps_excp_write`

Actually I found these points during the evolution of this PR. You can see from the test history there are many asserts and panics in the history version.
You can try to remove these two snippets, then you will meet the errors that  I have met.

tmedicci commented:
> Will try locally on ESP32-S2 to see whether can reproduce the issue @tmedicci
> 
> Update: have found the reason. ESP32-S2 is not like ESP32 and ESP32-S3, the former has `XCHAL_HAVE_LOOPS=0` in it's chip configuration, while the later have `XCHAL_HAVE_LOOPS=1`.
> 
> This difference triggered an error in a code path that have not been tested. Will fix it and update the patch.

Thanks, I'll trigger the CI again...

chirping78 commented:
> 
> I removed the clearing of the PS.EXCM in the stack and the results of the CI were the same!

I have added two `/* Overwrite the PS saved by the exception entry. */`,
one in `_xtensa_level1_handler`, and another in `_xtensa_syscall_handler`.

Which one are you in doubt? Or both?
I will try today to remove them to see whether there are any issue.

chirping78 commented:
Here is my test results today:

if removed the clearing of the `REG_PS.EXCM` from level1 handler, and from syscall handler (yes, both removed),
the ostest on esp32-devkitc, esp32s2-saola-1, esp32s3-devkit all passed.

But I have also tested on a hifi4 dsp, the result is:

removed the clearing of the `REG_PS.EXCM` from
 (1) level1 `yes`, syscall `no`, passed
 (2) level1 `yes`, syscall `yes`, failed. Unhandled Exception 2
 (3) level1 `no`, syscall `no`,  passed
 (4) level1 `no`, syscall `yes`, failed. Unhandled Exception 2

Please note that **both removed** looks right for esp32, but not for hifi4.

The failed scenario for hifi4, in case (2) and (4), is:
***a task was interrupted by `syscall`, then later was resumed by `RFI`***. This is the reason clearing of the `REG_PS.EXCM` can't be removed from syscall handler.

For the same reason, clearing of the `REG_PS.EXCM` can't be removed from level1 handler. Consider this scenario:
***a task was interrupted by level1 interrupt, then later was resumed by `RFI`***.
This scenario did not appear in the tests, but it does not mean that it will not happen in the future.
So the conclusion is clearing of the `REG_PS.EXCM` can not be removed from level1 and syscall handler.

tmedicci commented:
> if removed the clearing of the `REG_PS.EXCM` from level1 handler, and from syscall handler (yes, both removed), the ostest on esp32-devkitc, esp32s2-saola-1, esp32s3-devkit all passed.

I forget to mention that I have removed it from level1 handler.

> The failed scenario for hifi4, in case (2) and (4), is: _**a task was interrupted by `syscall`, then later was resumed by `RFI`**_. This is the reason clearing of the `REG_PS.EXCM` can't be removed from syscall handler.

> For the same reason, clearing of the `REG_PS.EXCM` can't be removed from level1 handler. Consider this scenario: _**a task was interrupted by level1 interrupt, then later was resumed by `RFI`**_. This scenario did not appear in the tests, but it does not mean that it will not happen in the future. So the conclusion is clearing of the `REG_PS.EXCM` can not be removed from level1 and syscall handler.

This is the point that I'm not convinced: in which situation a level-1 handler (or syscall handler) would be resumed by `RFI` instead of `RFE`? Isn't that an error (by design)?

chirping78 commented:
Take the task sleeping scenario as an example:
- task A call sleep(), this will finally trigger syscall handler;
- but when the syscall handler return, it will very likely return to idle task;
- With idle task being the active task, system will go through several timer interrupts, then task A sleep time expired, the timer interrupt will return to task A.

In above scenario, task A is interrrupted by `syscall` (a level1 interrupt), and if timer interrupt level in the system is 2 or 3, task A will be resumed by `RFI`

I think this is a common sense that in RTOS, the interrupt handle will not determinately return to the currently being interrupted task, isn't it?
https://github.com/apache/nuttx/blob/master/arch/xtensa/src/common/xtensa_user_handler.S#L289
https://github.com/apache/nuttx/blob/master/arch/xtensa/src/common/xtensa_int_handlers.S#L149

chirping78 commented:
@tmedicci any more comments?

tmedicci commented:
> @tmedicci any more comments?

Hi @chirping78 , I'm sorry for the delayed answer, I was on leave...

> Above information is enough to convince you

Yes, this is fine. Thanks and congratulations for the work. We will keep monitoring it with our internal CI.
